prov/util/src/util_mem_hooks.c

```diff
@@ -0,0 +1,572 @@
+{
+	struct ofi_patcher_dl_iter_context ctx = {
+		.patch    = patch,
+		.remove   = false,
+		.status   = FI_SUCCESS,
+	};
+
+	(void) dl_iterate_phdr(ofi_patcher_phdr_iterator, &ctx);
+
+	return ctx.status;
+}
+
+static int ofi_patcher_remove_patch(struct ofi_patcher_patch *patch)
+{
+	struct ofi_patcher_dl_iter_context ctx = {
+		.patch    = patch,
+		.remove   = true,
+		.status   = FI_SUCCESS,
+	};
+
+	/* Avoid locks here because we don't modify ELF data structures.
+	* Worst case the same symbol will be written more than once.
+	*/
+	(void) dl_iterate_phdr(ofi_patcher_phdr_iterator, &ctx);
+
+	return ctx.status;
+}
+
+static inline int ofi_patcher_get_orig(const char *symbol, void *replacement)
+{
+	const char *error;
+	void *func_ptr;
+
+	func_ptr = dlsym(RTLD_DEFAULT, symbol);
+	if (func_ptr == replacement) {
+		(void)dlerror();
+		func_ptr = dlsym(RTLD_NEXT, symbol);
+		if (func_ptr == NULL) {
+			error = dlerror();
+			/* FI_DBG(&core_prov, FI_LOG_MR, 
+			"get orig failed: %s\n", error); */
+			return -FI_EOVERFLOW;
+		}
+	}
+
+	return FI_SUCCESS;
+}
+
+int ofi_patcher_patch_symbol(const char *symbol_name, 
+			     uintptr_t replacement, uintptr_t *orig)
+{
+	int ret;
+	struct ofi_patcher_patch* patch = (struct ofi_patcher_patch *)
+					   malloc(sizeof(struct ofi_patcher_patch));
+	if (NULL == patch) {
+		return -FI_EOVERFLOW;
+	}
+	
+	patch->super.patch_symbol = strdup (symbol_name);
+	if (NULL == patch->super.patch_symbol) {
+		free(patch);
+		return -FI_EOVERFLOW;
+	}
+	
+	patch->super.patch_value = replacement;
+	patch->super.patch_restore = (void*) ofi_patcher_remove_patch;
+
+	/* Take lock first to handle a possible race where dlopen() is called
+	 * from another thread and we may end up not patching it.
+	 */
+	fastlock_acquire(&ofi_patcher_module.patch_list_mutex);
+	
+	ret = ofi_patcher_apply_patch(patch);
+	if (FI_SUCCESS != ret) {
+		goto unlock;
+	}
+
+	*orig = ofi_patcher_get_orig(patch->super.patch_symbol ,
+				    (void *) replacement);
+	if (FI_SUCCESS != ret) {
+		goto unlock;
+	}
+	dlist_insert_tail(&ofi_patcher_module.patch_list,
+			  &patch->super.super);
+
+
+unlock:
+	free(patch);
+	fastlock_release(&ofi_patcher_module.patch_list_mutex);
+
+	return ret;
+}
+
+/*SYS_MMAP, SYS_MUNMAP and etc. handling,
+*the main patcher initilization
+*/
+
+/* calling __syscall is preferred on some systems when some arguments may be 64-bit. it also
+ * has the benefit of having an off_t return type */
+//#define memory_patcher_syscall __syscall
+//#else
+#define memory_patcher_syscall syscall
+//#endif
+
+/*SYS_MMAP*/
+#if defined (SYS_mmap)
+
+static void *(*original_mmap)(void *, size_t, int, int, int, off_t);
+
+static void *_intercept_mmap(void *start, size_t length,
+			     int prot, int flags, int fd, off_t offset)
+{
+	void *result = 0;
+
+	/*not implemented yet*/
+	/*if ((flags & MAP_FIXED) && (start != NULL)) {
+		opal_mem_hooks_release_hook (start, length, true);
+	}*/
```

Comment:
The OMPI calls that intercept the memory calls (mmap, munmap, mremap, etc.), should all have a call to opal_mem_hooks_release_hook().  We will want to replace those with something like this:
```
ofi_patcher_handler(const void *addr, size_t len);
```
and define ofi_patcher_handler as:
```
void ofi_patcher_handler(const void *addr, size_t len)
{
    fastlock_acquire(&patcher.lock);
    ofi_monitor_notify(&patcher.monitor, addr, len);
    fastlock_release(&patcher.lock);
}
```
That will result in flushing any cached registrations that cover 'addr' + 'len'.